### PR TITLE
Add mapbox geocoding capability

### PR DESCRIPTION
**Pull request by attwad**
Even though it's a no-op this is quite a massive PR I'm sorry about that, I could have sent the mapbox/fake/index changes separately but having the context of the various commits might make more sense.

I will send the UI changes and corresponding integration tests that make use of the fake geocoder next.

I've documented a few things in the README in the geocoding folder I recommend looking at it first.

I've spent a few hours trying to mock things nicely for the mapbox.test.ts, I ended up with something that's not too terrible imo, let me know if you think there's a better way.

The constructor for the mapbox API (`Geocoding()` calls using the access token) will actually throw an error if the token is invalid, leading to the necessity of mocking all these calls in all unit tests calling the main.ts script.

### CODE REVIEW

File: verification/curator-service/api/src/geocoding/README.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+## Mapbox specificities
+
+You can play with the Mapbox geocoding API freely at [this URL](https://docs.mapbox.com/search-playground/).
+
+Check-out their [documentation](https://docs.mapbox.com/api/search/) as well.
+
+In order to be able to use the `mapbox.places-permanent` endpoint that is needed for storing the results as per Mapbox usage policy, you need to set the right secret token in the `MAPBOX_TOKEN` environment variable.
+
+If the `MAPBOX_PERMANENT_GEOCODE` environment variable isn't set, the geocoder will default to the `mapbox.places` endpoint instead which is useful during development.
```
Comment (khmoran) [inline verification/curator-service/api/src/geocoding/README.md:32]: this is a boolean? WDYT about making it MAPBOX_ENDPOINT and setting it to 'mapbox.places' or 'mapbox.places-permanent', just to make it super explicit?

fine with it as is though as long as I'm understanding it correctly

File: verification/curator-service/api/test/geocoding/fake.test.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
+import app from '../../src/index';
```
Comment (khmoran) [inline verification/curator-service/api/test/geocoding/fake.test.ts:1]: you are truly impressive, sir, that you wrote a test for your fake. I applaud you!

File: verification/curator-service/api/test/index.test.ts [hunk 3]
```diff
@@ -19,6 +20,10 @@ afterAll(() => {
     return mongoose.disconnect();
 });
 
+beforeEach(() => {
```
Comment (khmoran) [inline verification/curator-service/api/test/index.test.ts:23]: I feel like this addition is the result of some sadness and pain and I want you to know that you're seen

File: verification/curator-service/api/src/geocoding/README.md (hunk 1 above)
Comment (attwad) [reply verification/curator-service/api/src/geocoding/README.md:32]: ah good point, the env validatation library we're using supports a bunch of various other types so yeah I changed to use those :) thanks!

File: verification/curator-service/api/src/geocoding/fake.ts [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+export default class FakeGeocoder {
+    private entries: Map<string, GeocodeResult>;
+
+    constructor() {
+        this.entries = new Map<string, GeocodeResult>();
+    }
+    async geocode(query: string): Promise<GeocodeResult[]> {
+        const entry = this.entries.get(query);
+        if (entry) {
```
[the remark this answers is not in the document]
Comment (attwad) [reply verification/curator-service/api/src/geocoding/fake.ts:13]: cheers, done

File: verification/curator-service/api/src/geocoding/geocoder.ts [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+export interface GeocodeResult {
+    geometry: {
+        latitude: number;
+        longitude: number;
+    };
+    country: string;
+    // First administrative division (regions in the US, Länder in Germany, ...).
+    administrativeAreaLevel1: string | undefined;
+    // Second administrative division (county in the US, departments in France, ...).
+    administrativeAreaLevel2: string | undefined;
```
[the remark this answers is not in the document]
Comment (attwad) [reply verification/curator-service/api/src/geocoding/geocoder.ts:11]: Interesting, I've made a few comments to the CL, I'll work on the UI part of things today to not conflict and will amend this CL after you've submitted yours, you got priority based on the time you have this week :)

File: verification/curator-service/api/src/geocoding/mapbox.ts [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+                    query: query,
+                    language: ['en'],
+                    limit: 5,
+                })
+                .send();
+            const features = (resp.body as GeocodeResponse).features;
+            if (features.length == 0) {
+                return [];
+            }
+            return features.map((feature) => {
```
[the remark this answers is not in the document]
Comment (attwad) [reply verification/curator-service/api/src/geocoding/mapbox.ts:42]: ah yeah, I was throwing an error here before hence this if() but it's not needed anymore and forgot to remove it, thanks

File: verification/curator-service/api/src/util/validate-env.ts [hunk 7]
```diff
@@ -14,6 +14,8 @@ export default function validateEnv(): Readonly<{
     AFTER_LOGIN_REDIRECT_URL: string;
     STATIC_DIR: string;
     ENABLE_LOCAL_AUTH: string;
+    MAPBOX_TOKEN: string;
+    MAPBOX_PERMANENT_GEOCODE: string;
```
[the remark this answers is not in the document]
Comment (attwad) [reply verification/curator-service/api/src/util/validate-env.ts]: there is! done.

File: verification/curator-service/api/test/geocoding/fake.test.ts (hunk 2 above)
Comment (attwad) [reply verification/curator-service/api/test/geocoding/fake.test.ts:1]: I am only doing my duty. :)

File: verification/curator-service/api/test/index.test.ts (hunk 3 above)
Comment (attwad) [reply verification/curator-service/api/test/index.test.ts:23]: 👀

File: verification/curator-service/api/src/geocoding/geocoder.ts [hunk 8]
```diff
@@ -0,0 +1,20 @@
+// A single result from a Geocoder.
+export interface GeocodeResult {
+    geometry: {
+        latitude: number;
+        longitude: number;
+    };
+    country: string;
+    // First administrative division (regions in the US, Länder in Germany, ...).
```
Comment (allysonjp715) [inline verification/curator-service/api/src/geocoding/geocoder.ts:8]: nit: is it "states" in the US?

File: verification/curator-service/api/src/geocoding/geocoder.ts [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+    administrativeAreaLevel1: string | undefined;
+    // Second administrative division (county in the US, departments in France, ...).
+    administrativeAreaLevel2: string | undefined;
+    locality: string | undefined;
```
Comment (allysonjp715) [inline verification/curator-service/api/src/geocoding/geocoder.ts:12]: Could you add a comment for locality too?

File: verification/curator-service/api/src/geocoding/mapbox.ts [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+    GeocodeFeature,
+    GeocodeMode,
+    GeocodeResponse,
+    GeocodeService,
+} from '@mapbox/mapbox-sdk/services/geocoding';
+
+import { GeocodeResult } from './geocoder';
+import { MapiResponse } from '@mapbox/mapbox-sdk/lib/classes/mapi-response';
+
+function getFeatureTypeFromContext(
```
Comment (allysonjp715) [inline verification/curator-service/api/src/geocoding/mapbox.ts:11]: Could you add a comment to this function? I'm not following what it does.

And is there a reason it's outside of the class if it isn't exported?

**Comment by attwad** [comment]
[Implement geocoding of new cases using mapbox API](https://trello.com/c/K6SeR1oF/218-implement-geocoding-of-new-cases-using-mapbox-api)

File: verification/curator-service/api/test/geocoding/fake.test.ts [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+    it('can seed geocodes', (done) => {
+        request(app)
+            .post('/api/geocode/seed')
+            .send({
+                administrativeAreaLevel1: 'Rhône',
+                country: 'France',
+                geometry: { latitude: 45.75889, longitude: 4.84139 },
+                text: 'Lyon',
+            })
+            .expect(200, done);
```
Comment (allysonjp715) [inline verification/curator-service/api/test/geocoding/fake.test.ts:34]: Could check here that you can query for the geocode that's been seeded

File: verification/curator-service/api/src/geocoding/geocoder.ts (hunk 8 above)
Comment (attwad) [reply verification/curator-service/api/src/geocoding/geocoder.ts:8]: right sorry :) fixed.

File: verification/curator-service/api/src/geocoding/geocoder.ts (hunk 9 above)
Comment (attwad) [reply verification/curator-service/api/src/geocoding/geocoder.ts:12]: this is going away as per Kelly's refactoring of other fields, I'll align those once she submits her other PR.

File: verification/curator-service/api/src/geocoding/mapbox.ts (hunk 10 above)
Comment (attwad) [reply verification/curator-service/api/src/geocoding/mapbox.ts:11]: Added a comment.
It has no need for anything in the class, I could make it a static method but I see no real benefits in doing so.

File: verification/curator-service/api/test/geocoding/fake.test.ts (hunk 11 above)
Comment (attwad) [reply verification/curator-service/api/test/geocoding/fake.test.ts:34]: I am not exposing an endpoint to query the geocodes yes (doing that in a later PR), I will add that when it is exposed.

**Pull request merged**